crewAIInc/crewAI
# [Bug Fix] -  Agent Delegation Tools Not Being Updated

### Problem
When an agent is marked with `allow_delegation=True` and have other tools assigned to it, it cannot access delegation tools. This occurs because the current implementation of `_add_delegation_tools` only updates `task.tools` but not `agent.tools`, causing the agent to lose access to delegation capabilities.

### Solution
Updated `_add_delegation_tools` to properly manage both `task.tools` and `agent.tools`, ensuring delegation tools are correctly assigned and accessible.

### Minimal Reproducible Example

```python
from crewai import Agent, Crew, Task

# Create a simple tool
class SimpleTool:
    name = "simple_tool"
    def __call__(self): pass

# Setup agents
master_agent = Agent(
    name="master",
    tools=[SimpleTool()],
    allow_delegation=True  # This agent should be able to delegate
)

worker_agent = Agent(
    name="worker",
    tools=[SimpleTool()]
)

# Create crew
crew = Crew(
    agents=[master_agent, worker_agent],
    tasks=[Task(
        description="Test task",
        agent=master_agent
    )]
)

# Before fix: master_agent cannot access delegation tools
# After fix: master_agent can properly access delegation tools
```

### Details
- The bug was in `crew.py`'s `_add_delegation_tools` method
- Previously, delegation tools were only added to `task.tools`
- Fixed by introducing `add_or_update_tools` helper function to manage tool lists
- Now updates both `agent.tools` and `task.tools`

### Testing
To verify the fix:
1. Create a crew with multiple agents
2. Set `allow_delegation=True` on the master agent
3. Confirm the master agent can access and use delegation tools

### Breaking Changes
None. This is a bug fix that maintains existing API compatibility.
